### PR TITLE
[e2e]: Add external E2E tests

### LINKED ISSUE #203: [build] Create release version container images
**Is this a BUG REPORT or FEATURE REQUEST?**:
feature

**What happened**:
We currently provide commit based container images.

**What you expected to happen**:
Provide per release images such as `c.txtdirect.org/txtdirect:v0.4.0` on tags.

### LINKED ISSUE #291: [tests]: Discuss external E2E tests
**Is this a BUG REPORT or FEATURE REQUEST?**:
feature

**What happened**:
Currently, our E2E tests don't cover every part of the project and we need to have tests for more complex scenarios to be aware of the edge cases.

**What you expected to happen**:
Discuss how we're going to implement these tests and how the structure is supposed to be.

I think it would be good to have an isolated and fresh environment for each test case and keep all the records related to a test case inside its own zone file.
We can separate the test cases by keeping them in their own directories and have a manager script at the root that runs each test cases' script.
The structure would look like this:
```
├─e2e_tests
├── dockerv2
│   ├── pull
│   │   ├── zonefile
│   │   └── test.sh
│   └── push
│       ├── zonefile
│       └── test.sh
└── manager.sh
```

Not sure how we should handle the containers for each test case. We can use solutions like `compose` or have our own script for creating and managing the test containers.
I personally prefer the custom script and a downside would be requiring more time to implement a stable solution.
We can also add the container-related stuff to each test's script but think it would be better to keep the test script as clear and simple as possible and only keep the tests inside it.

Would love to hear other thoughts and approaches.

**How to reproduce it (as minimally and precisely as possible)**:
Feature Request

**Anything else we need to know?**:
Questions:
- A better structure for managing the files?
- Alternatives for compose and custom script?
- Other solutions for managing the test cases instead of a manager bash script?

### PR DESCRIPTION
**Pull request by erbesharat**
**What this PR does / why we need it**:
Adds external e2e tests

**Which issue this PR fixes**:
fixes #291 
fixes #203 

**Special notes for your reviewer**:
Covered types:

- [x] `host`
- [x] `path`
- [x] `gometa`
- [ ] `dockerv2`
- [x] `git`
- [ ] `proxy`

Questions:

- What should we do about adding the domains to /etc/hosts? Should we have a list of URIs in /e2e and write it to `/etc/hosts` when it gets changed? That way whenever we want to add a record to a test-case we have to add it to /e2e/URIs so the manager-script can read it and add them to /etc/hosts.

### CODE REVIEW

**Comment by erbesharat** [comment]
We should reconsider our flow for this test: 
https://github.com/txtdirect/txtdirect/blob/2164d93aeae9f79560508236f5c752a99f9c1973/e2e/host/main.go#L44-L51

Right now since we don't return an error in `getBaseTarget` when `to=` is empty the test I mentioned gets redirected to itself.

Shouldn't we return an error in here so the fallback can get triggered?

https://github.com/txtdirect/txtdirect/blob/ec59e2cdbcf4fe72cf20b3ac2a8f20c114de1fc5/txtdirect.go#L53-L62

**Comment by erbesharat** [comment]
Currently, the old fallback tests fail and there is a permission issue with Travis. Will migrate the fallback tests at the end after we have tests for all the types.

File: e2e/main.go [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+		EndpointsConfig: map[string]*network.EndpointSettings{
+			"coretxtd": &network.EndpointSettings{NetworkID: d.network.ID},
+		},
+	}, "")
+	if err != nil {
+		return fmt.Errorf("Couldn't create the CoreDNS container: %s", err.Error())
+	}
+
+	d.txtdContainer, err = d.cli.ContainerCreate(d.ctx, &container.Config{
+		Image: "okkur/txtdirect:0.4.0",
```
Comment (stp-ip) [inline e2e/main.go]: We could use `c.txtdirect.org/txtdirect:0.4.0-dirty` here. Also the version of the container should probably be derived from the makefile or else the build container might get out of sync with using it here. What do you think?

File: Makefile [hunk 2]
```diff
@@ -32,5 +32,10 @@ docker-build:
 	docker run --network=host -v $(shell pwd):/source -v $(GOPATH)/pkg/mod:/go/pkg/mod golang:1.13-alpine /bin/sh \
 	-c "cd /source && apk add git gcc musl-dev make && make build"
 
+endtoend-test:
+	cd e2e && \
+	docker build -t c.txtdirect.org/tester:0.0.1 . && \
```
Comment (stp-ip) [inline Makefile]: What about using `c.txtdirect.org/tester:0.0.1-dirty` to make clear this is a build from code, that's not meant to be used anywhere?
Also do we want to version this and if yes how?

**Comment by stp-ip** [comment]
Additional notes:

github.com/docker/docker should be replaced with github.com/docker/engine within go.mod as far as I can see. Currently it works as we are using @master for the version instead of a tagged release, which would not be available within github.com/docker/docker.

make docker-build should use `--network=host`

Mounting the pkg directory as we do in our other docker run usage would speed up the tests.

ContainerCreate doesn't pull any container images. So it assumes that `coredns/coredns` for example is already present on the host. Either we do this within the makefile or the main.go file needs to pull in the specific container image before creating the container.

File: e2e/host/main.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		},
+		expected: "https://noversion.host.host.example.com",
+	},
+	{
+		name: "Redirect to a host record's unspecified to=",
+		args: data{
+			host: "noto.host.host.example.com",
+			path: "/",
+		},
+		expected: "http://noto.host.host.example.com/",
```
[the remark this answers is not in the document]
Comment (erbesharat) [reply e2e/host/main.go]: As I mentioned in my first comment, we should reconsider this test (it was in our old e2e tests) and return an error and trigger the fallback when `to=` is empty.

File: e2e/main.go [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+)
+
+var resultRegex = regexp.MustCompile("Total:+\\s(\\d+),\\sPassed:+\\s(\\d+),\\sFailed:+\\s(\\d+)")
```
[the remark this answers is not in the document]
Comment (erbesharat) [reply e2e/main.go]: After the path issues I've fallen in love with Regex :D It's really awesome.

File: e2e/main.go (hunk 1 above)
Comment (erbesharat) [reply e2e/main.go]: Yeah, it should definitely read the version from somewhere else like Makefile as you mentioned and not hardcode it. We can set an env variable in `endtoend-test` target for the image name or version and use `os.Getenv()` in here instead.

File: Makefile (hunk 2 above)
Comment (erbesharat) [reply Makefile]: I don't think versioning it would be necessary. Should we use something like `c.txtdirect.org/tester:dirty` or `c.txtdirect.org/tester` is enough?

File: e2e/host/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (stp-ip) [reply e2e/host/main.go]: Shouldn't this be 404 in general with no other fallback configured?

File: Makefile (hunk 2 above)
Comment (stp-ip) [reply Makefile]: `c.txtdirect.org/tester:dirty` sounds good.

File: e2e/main.go [hunk 5]
```diff
@@ -337,3 +370,15 @@ func (d *dockerManager) CountStats(stats []string) error {
 
 	return nil
 }
+
+func (d *dockerManager) PullImages() error {
+	_, err := d.cli.ImagePull(d.ctx, corednsImage, types.ImagePullOptions{})
+	if err != nil {
+		return fmt.Errorf("Coudln't pull the CoreDNS image: %s", err.Error())
```
Comment (stp-ip) [inline e2e/main.go]:
```suggestion
		return fmt.Errorf("Couldn't pull the CoreDNS image: %s", err.Error())
```

File: e2e/main.go [hunk 6]
```diff
@@ -337,3 +370,15 @@ func (d *dockerManager) CountStats(stats []string) error {
...
+	}
+	_, err = d.cli.ImagePull(d.ctx, txtdirectImage, types.ImagePullOptions{})
+	if err != nil {
+		return fmt.Errorf("Coudln't pull the TXTDirect image: %s", err.Error())
```
Comment (stp-ip) [inline e2e/main.go]:
```suggestion
		return fmt.Errorf("Couldn't pull the TXTDirect image: %s", err.Error())
```

**Comment by erbesharat** [comment]
Sorry for the force pushes :grin: Didn't want to make more debugging commits.

File: e2e/gometa/zonefile [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+                                      2017101003   ; serial
+                                      5m           ; refresh
+                                      5m           ; retry
+                                      1w           ; expire
+                                      12h    )     ; minimum
+
+@                  86400 IN NS      ns.example.net.
+@                  86400 IN NS      ns.example.nl.
+
+$TTL 1H
```
[the remark this answers is not in the document]
Comment (stp-ip) [reply e2e/gometa/zonefile]: Still valid ;)

File: e2e/host/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (stp-ip) [reply e2e/host/main.go]: ?

[the remark this answers is not in the document]
Comment (erbesharat) [reply e2e/host/main.go]: https://github.com/txtdirect/txtdirect/blob/97a881d2486ed2676768b9bb6c7845e1ec138cc5/record.go#L192
I thought this would take care of it but it only gets called when `to=` exist. Will add a condition to trigger fallback when the type is `host` and `to=` doesn't exist.

[the remark this answers is not in the document]
Comment (stp-ip) [reply e2e/host/main.go]: +1
